### ISSUE TITLE
E2E plugin requires Android SDK 24

### ISSUE THREAD

**Issue by alextekartik** [flutter/flutter#48701]
Although [E2E](https://pub.dev/packages/e2e) plugin has declared `minSdkVersion` set to 16 in its gradle file, it uses `CompletableFuture` only defined in Android SDK 24, causing a crash upon registration on device with API lower than 24.

```
01-13 12:24:38.056 22328 22328 D AndroidRuntime: Shutting down VM
01-13 12:24:38.061 22328 22328 E AndroidRuntime: FATAL EXCEPTION: main
01-13 12:24:38.061 22328 22328 E AndroidRuntime: Process: com.terkartik.sqfliteexample, PID: 22328
01-13 12:24:38.061 22328 22328 E AndroidRuntime: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/util/concurrent/CompletableFuture;
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at dev.flutter.plugins.e2e.E2EPlugin.<clinit>(E2EPlugin.java:22)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at io.flutter.plugins.GeneratedPluginRegistrant.registerWith(GeneratedPluginRegistrant.java:15)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at com.tekartik.sqfliteexample.MainActivity.onCreate(MainActivity.java:12)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:6362)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1108)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2441)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2548)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at android.app.ActivityThread.access$1100(ActivityThread.java:154)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1399)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:102)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:157)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:5609)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:771)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:649)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: Caused by: java.lang.ClassNotFoundException: Didn't find class "java.util.concurrent.CompletableFuture" on path: DexPathList[[zip file "/data/app/com.terkartik.sqfliteexample-1/base.apk"],nativeLibraryDirectories=[/data/app/com.terkartik.sqfliteexample-1/lib/arm64, /data/app/com.terkartik.sqfliteexample-1/base.apk!/lib/arm64-v8a, /vendor/lib64, /system/lib64]]
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	... 15 more
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	Suppressed: java.lang.ClassNotFoundException: java.util.concurrent.CompletableFuture
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 		at java.lang.Class.classForName(Native Method)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 		at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 		at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 		at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 		... 16 more
01-13 12:24:38.061 22328 22328 E AndroidRuntime: 	Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available
```
Solutions:
- Explicitely set minSdkTarget to 24, although it will prevent some application from testing on lower Android devices (not preferred solution)
- Support Android 16 by changing the java code: https://github.com/flutter/plugins/blob/master/packages/e2e/android/src/main/java/dev/flutter/plugins/e2e/E2EPlugin.java

## Steps to Reproduce

1. Run the E2E example app on an android emulator 5.1 (API less than 24).
2. Observe the native crash using `logcat`

**Expected results:** 

No crash

**Actual results:** 

Crash, native stack pasted above

**Comment by dnfield** [comment]
We definitely want to support this on lower than 24 - given the purpose of this package.

I've put up a PR that provides a different, compatible implementation of CompletableFuture.

**Comment by dnfield** [comment]
This will be published as v0.4.3 as soon as I can figure out what permissions I need to do that :)

**Comment by MaikuB** [comment]
@dnfield apologies for replying to an old thread but is it possible that in one of the next e2e updates that the changelog could be retroactively updated to mention it fixes a crash around this. Had a user of my plugin report this issue but it never occurred to me that e2e was causing the crash as the changelog entry read more like it was improving device compatibility. Happy to look at submitting a PR for this myself if that helps.
